### DISCUSSION TITLE
How to Retrieve User Locale from Microsoft Teams Context in a React Tab App?

### DISCUSSION THREAD

**Discussion by sayan-mondal-dcs** [OfficeDev/microsoft-365-agents-toolkit#12242] [Q&A]
Hello everyone,

I'm working on a React-based Microsoft Teams tab application where I need to dynamically set the user interface language based on the user's locale settings within Microsoft Teams. I've been able to manage user preferences and locale settings but I want to fetch the locale directly from the Teams context to ensure it aligns with the user's Teams settings.

My Current Setup:
Framework: React with TypeScript
Microsoft Teams SDK: @microsoft/teams-js
Context Management: I'm using React.createContext to manage the theme and other settings.
Localization: Currently handling localization using react-i18next and managing supported languages like en-us, hin-in, and it-it.
What I Want to Achieve:
I want to:

Retrieve the user's locale directly from the Teams context.
Use this locale to adjust the UI language dynamically.
Possibly update the global context I'm using (TeamsFxContext) to include this locale for consistent use across the application.
What I've Tried So Far:
I’ve attempted to use the microsoftTeams.getContext() method, but I'm looking for best practices or examples on how to integrate this properly within my React components, especially in the context of a larger application.

Sample Code:
Here’s a simplified version of my current approach:
```import { useGraphWithCredential, useTeamsUserCredential } from "@microsoft/teamsfx-react";
import { useTranslation } from "react-i18next";
import { useEffect, useState } from "react";
import { Providers, ProviderState } from "@microsoft/mgt-element";
import { TeamsFxProvider } from "@microsoft/mgt-teamsfx-provider";

export default function App() {
  const [userPreferredLanguage, setUserPreferredLanguage] = useState('en-us'); // default language
  const supportedLanguages = ['en-us', 'hin-in', 'it-it'];

  const { t, i18n } = useTranslation();

  useEffect(() => {
    const fetchPreferredLanguage = async () => {
      setUserPreferredLanguage('en-us'); // fallback to 'en-us' if no preference found
    };
    fetchPreferredLanguage();
  }, [userPreference]);

  useEffect(() => {
    i18n.changeLanguage(userPreferredLanguage);
  }, [i18n, userPreferredLanguage]);

  return (
    <TeamsFxContext.Provider value={{ theme, themeString, teamsUserCredential, locale: userPreferredLanguage }}>
         ///
    </TeamsFxContext.Provider>
  );
}
```
References:
[How to get MS Teams (Desktop App) language programmatically?](https://stackoverflow.com/questions/71967840/how-to-get-ms-teams-desktop-app-language-programmatically)](https://stackoverflow.com/questions/71967840/how-to-get-ms-teams-desktop-app-language-programmatically)

[How to set time zone in microsoft teams](https://answers.microsoft.com/en-us/msteams/forum/all/how-to-set-time-zone-in-microsoft-teams/ce83d312-6e86-4ee3-bdd2-0abda9d303f7)

Questions:
I'm not able to get the locale from Microsoft Teams React tab app?
Are there any pitfalls or improvements I should consider when using microsoftTeams.getContext() for this purpose?
How can I ensure that the locale is correctly passed down through my context and available throughout the app?
Additional Context:
I’m also using react-i18next for translation management.
The application is integrated with Microsoft Teams via Teams FX, and I'm also leveraging the Microsoft Graph API.
Any guidance would be greatly appreciated!
Thank you!

**Comment by VikrantSingh01** [comment]
### Retrieving User Locale from Teams Context in a React Tab App

In Teams JS SDK v2, the old callback-based `microsoftTeams.getContext(callback)` API has been replaced with a Promise-based approach using the `app` namespace. The locale is now available at `context.app.locale` (instead of the flat `context.locale` from v1).

### Step 1: Initialize and get context

You **must** call `app.initialize()` before calling `app.getContext()`. Both return Promises:

```typescript
import { app } from "@microsoft/teams-js";

await app.initialize();
const context: app.Context = await app.getContext();
const userLocale: string = context.app.locale; // e.g. "en-us"
```

The `context.app.locale` property returns the user's current locale formatted as `languageId-countryId` (for example, `en-us`, `hi-in`, `it-it`).

### Step 2: Create a custom hook

This encapsulates the initialization and context retrieval cleanly:

```typescript
import { useState, useEffect } from "react";
import { app } from "@microsoft/teams-js";

export function useTeamsLocale(fallback: string = "en-us") {
  const [locale, setLocale] = useState<string>(fallback);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    let cancelled = false;

    async function initAndGetLocale() {
      try {
        await app.initialize();
        const context = await app.getContext();
        if (!cancelled) {
          setLocale(context.app.locale ?? fallback);
        }
      } catch (err) {
        console.error("Failed to get Teams context:", err);
      } finally {
        if (!cancelled) {
          setIsLoading(false);
        }
      }
    }

    initAndGetLocale();
    return () => { cancelled = true; };
  }, [fallback]);

  return { locale, isLoading };
}
```

### Step 3: Wire it together with react-i18next

```tsx
import React, { useEffect } from "react";
import { useTranslation } from "react-i18next";
import { useTeamsLocale } from "./useTeamsLocale";

const SUPPORTED_LANGUAGES = ["en", "hi", "it"];

function App() {
  const { i18n, t } = useTranslation();
  const { locale, isLoading } = useTeamsLocale("en-us");

  useEffect(() => {
    // context.app.locale returns "en-us", "hi-in", etc.
    // Extract the language prefix to match your i18n resource keys
    const langPrefix = locale.split("-")[0];
    const language = SUPPORTED_LANGUAGES.includes(langPrefix) ? langPrefix : "en";
    i18n.changeLanguage(language);
  }, [locale, i18n]);

  if (isLoading) return <div>Loading...</div>;

  return (
    <div>
      <h1>{t("welcome")}</h1>
      <p>Current locale: {locale}</p>
    </div>
  );
}

export default App;
```

### Key points to keep in mind

- **`app.initialize()` must be called first.** The SDK will throw `"The library has not yet been initialized"` if you call `app.getContext()` before initialization completes.

- **The v1 `microsoftTeams.getContext(callback)` is deprecated.** In v2, use `app.getContext()` which returns a `Promise<app.Context>`. The locale moved from the flat `context.locale` to the nested `context.app.locale`.

- **`@microsoft/teamsfx-react` and `useTeamsUserCredential` are deprecated.** The TeamsFx SDK is in deprecation mode (community-only support until September 2026). Use `@microsoft/teams-js` directly for context retrieval instead.

- **`context.app.osLocaleInfo` provides additional detail.** If you need OS-level formatting info (date/time formats, regional settings), check `context.app.osLocaleInfo` which returns a `LocaleInfo` object with properties like `shortDate`, `longDate`, `shortTime`, `longTime`, and `timezone`.

### References

- [Get context for your tab (Microsoft Learn)](https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/access-teams-context)
- [`app` module API reference](https://learn.microsoft.com/en-us/javascript/api/@microsoft/teams-js/app)
- [TeamsJS v1 to v2 migration guide](https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/using-teams-client-library)
